File: src/main/java/at/rueckgr/spotlight/service/DownloadDirectoryService.java
package at.rueckgr.spotlight.service;

import at.rueckgr.spotlight.model.DownloadableImage;
import at.rueckgr.spotlight.model.DownloadedImage;
import at.rueckgr.spotlight.util.SpotlightException;
import lombok.RequiredArgsConstructor;
import lombok.extern.log4j.Log4j2;
import org.apache.commons.codec.digest.DigestUtils;
import org.apache.commons.io.FilenameUtils;
import org.apache.commons.io.IOUtils;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Log4j2
@RequiredArgsConstructor
public class DownloadDirectoryService {

    public static final String FOLDER_NAME = "images";
    public static final String PRETTY_FILENAME_PATTERN = "^((?!\\W|^R(.){5,6}$).|,| | |-|\\.|\\(|\\)|')*$";

    private static final List<String> preferredLanguages = Arrays.asList("en", "de");

    private final Map<String, DownloadedImage> downloadedImages = new HashMap<>();

    private final ImageDownloaderService imageDownloaderService = new ImageDownloaderService();
    private final MetadataService metadataService;

    public void init() {
        createDirectory();
        metadataService.loadMetadata();
        loadDownloadedImages();
    }

    private void createDirectory() {
        File directory = new File(FOLDER_NAME);
        if (!directory.exists()) {
            if(!directory.mkdir()) {
                throw new SpotlightException("Unable to create download directory");
            }
        }
    }

    public void loadDownloadedImages() {
        final File directory = new File(FOLDER_NAME);
        final File[] directoryListing = directory.listFiles();
        if (directoryListing != null) {
            for (File file : directoryListing) {
                if (FilenameUtils.getExtension(file.getName()).equalsIgnoreCase("jpg")) {
                    try {
                        final String hash = DigestUtils.sha256Hex(IOUtils.toByteArray(new FileInputStream(file)));
                        downloadedImages.put(hash, new DownloadedImage(file, hash, metadataService.getLocale(hash)));
                    }
                    catch (IOException e) {
                        throw new SpotlightException(e);
                    }
                }
            }
        }
    }

    public synchronized void saveImageIfNotExists(final DownloadableImage downloadableImage) {
        final String hash = downloadableImage.getHash();
        if (downloadedImages.containsKey(hash)) {
            log.info("File already exists: {}", downloadableImage.toString());
            log.info("Existing file: {}", downloadedImages.get(hash));
            renameImageIfRequired(downloadableImage);
            return;
        }

        byte[] newImageBytes = imageDownloaderService.fetchImage(downloadableImage);
        final File nextFile = getNextFile(downloadableImage.getDescription());
        try {
            Files.copy(new ByteArrayInputStream(newImageBytes), nextFile.toPath());
        }
        catch (IOException e) {
            throw new SpotlightException(e);
        }
        downloadedImages.put(hash, new DownloadedImage(nextFile, hash, downloadableImage.getDownloadUrl().getLocale()));

        log.info(String.format("New image added : %s", downloadableImage));
    }

    private void renameImageIfRequired(final DownloadableImage downloadableImage) {
        final String hash = downloadableImage.getHash();
        final DownloadedImage downloadedImage = downloadedImages.get(hash);

        final boolean oldFileNeedRename =  !FilenameUtils.removeExtension(downloadedImage.getName()).matches(PRETTY_FILENAME_PATTERN);
        final String newFilename = downloadableImage.getDescription();
        final boolean newFileIsValidName = newFilename.matches(PRETTY_FILENAME_PATTERN);

        final String oldLanguage = downloadedImage.getLocale().getLanguage();
        final String newLanguage = downloadableImage.getDownloadUrl().getLocale().getLanguage();
        final boolean betterLanguageFound = (!preferredLanguages.contains(oldLanguage) && preferredLanguages.contains(newLanguage)) ||
                (preferredLanguages.contains(oldLanguage) && preferredLanguages.contains(newLanguage) && preferredLanguages.indexOf(oldLanguage) < preferredLanguages.indexOf(newFilename));

        log.debug("oldFileNeedRename: {}, betterLanguageFound: {}, newFileIsValidName: {}", oldFileNeedRename, betterLanguageFound, newFileIsValidName);
        if ((oldFileNeedRename || betterLanguageFound) && newFileIsValidName) {
            final File newFile = getNextFile(newFilename);
            if (!downloadedImage.getFile().renameTo(newFile)) {
                throw new SpotlightException(String.format("Unable to rename file %s to %s", downloadedImage.getName(), newFile.getName()));
            }

            log.info("Renaming {} to {}", downloadedImage.getName(), newFile.getName());
            downloadedImages.put(hash, new DownloadedImage(newFile, hash, downloadableImage.getDownloadUrl().getLocale()));
        }
    }

    private File getNextFile(final String filename) {
        File file = new File(FOLDER_NAME + File.separator + filename + ".jpg");
        int fileNo = 0;
        while (file.exists()) {
            fileNo++;
            file = new File(FOLDER_NAME + File.separator + filename + fileNo + ".jpg");
        }
        return file;
    }

    public String getFolderName() {
        return FOLDER_NAME;
    }

    public Map<String, DownloadedImage> getDownloadedFiles() {
        return this.downloadedImages;
    }
}

File: src/main/kotlin/at/rueckgr/spotlight/service/DownloadDirectoryService.kt
package at.rueckgr.spotlight.service

import at.rueckgr.spotlight.model.DownloadableImage
import at.rueckgr.spotlight.model.DownloadedImage
import at.rueckgr.spotlight.util.Logging
import at.rueckgr.spotlight.util.SpotlightException
import at.rueckgr.spotlight.util.logger
import org.apache.commons.codec.digest.DigestUtils
import org.apache.commons.io.FilenameUtils
import org.apache.commons.io.IOUtils
import java.io.ByteArrayInputStream
import java.io.File
import java.io.FileInputStream
import java.nio.file.Files

class DownloadDirectoryService (private val metadataService: MetadataService, private val metricsService: MetricsService) : Logging {
    val folderName = "images"
    private val prettyFilenamePattern = "^((?!\\W|^R(.){5,6}$).|,| | |-|\\.|\\(|\\)|')*$".toRegex()
    private val preferredLanguages = arrayOf("en", "de")

    val downloadedImages = HashMap<String, DownloadedImage>()

    private val imageDownloaderService = ImageDownloaderService()

    fun init() {
        createDirectory()
        metadataService.loadMetadata()
        loadDownloadedImages()
    }

    private fun createDirectory() {
        val directory = File(folderName)
        if (!directory.exists()) {
            if (!directory.mkdir()) {
                throw SpotlightException("Unable to create download directory")
            }
        }
    }

    private fun loadDownloadedImages() {
        val directory = File(folderName)
        val directoryListing = directory.listFiles()
        if (directoryListing != null) {
            for(file in directoryListing) {
                if (FilenameUtils.getExtension(file.name).equals("jpg", ignoreCase = true)) {
                    val hash = DigestUtils.sha256Hex(IOUtils.toByteArray(FileInputStream(file)))
                    downloadedImages[hash] = DownloadedImage(file, hash, metadataService.getLocale(hash))
                }
            }
        }
    }

    fun saveImageIfNotExists(downloadableImage: DownloadableImage) {
        val hash = downloadableImage.hash
        if (downloadedImages.containsKey(hash)) {
            logger().info("File already exists: {}", downloadableImage)
            logger().info("Existing file: {}", downloadedImages[hash])
            renameImageIfRequired(downloadableImage)
            return
        }

        val newImageBytes = imageDownloaderService.fetchImage(downloadableImage)
        val nextFile = getNextFile(downloadableImage.description)
        Files.copy(ByteArrayInputStream(newImageBytes), nextFile.toPath())
        downloadedImages[hash] = DownloadedImage(nextFile, hash, downloadableImage.downloadUrl.locale)

        logger().info("New image added: {}", downloadableImage)
        metricsService.recordMetric(MetricsService.Metric.DOWNLOAD_IMAGE)
    }

    private fun renameImageIfRequired(downloadableImage: DownloadableImage) {
        if (isRenameRequired(downloadableImage)) {
            val newFile = getNextFile(downloadableImage.description)
            val hash = downloadableImage.hash
            val downloadedImage = downloadedImages[hash]
            if (!downloadedImage!!.file.renameTo(newFile)) {
                throw SpotlightException(String.format("Unable to rename file %s to %s", downloadedImage.name, newFile.name))
            }

            logger().info("Renaming {} to {}", downloadedImage.name, newFile.name)
            downloadedImages[hash] = DownloadedImage(newFile, hash, downloadableImage.downloadUrl.locale)

            metricsService.recordMetric(MetricsService.Metric.RENAME_IMAGE)
        }
    }

    private fun isRenameRequired(downloadableImage: DownloadableImage): Boolean {
        val downloadedImage = downloadedImages[downloadableImage.hash]

        val oldFileNeedRename = !FilenameUtils.removeExtension(downloadedImage!!.name).matches(prettyFilenamePattern)
        val newFilename = downloadableImage.description
        val newFileIsValidName = newFilename.matches(prettyFilenamePattern)

        val oldLanguage = downloadedImage.locale?.language ?: ""
        val newLanguage = downloadableImage.downloadUrl.locale.language
        val betterLanguageFound = (oldLanguage !in preferredLanguages && newLanguage in preferredLanguages) ||
                (oldLanguage in preferredLanguages && preferredLanguages.contains(newLanguage) &&
                        preferredLanguages.indexOf(oldLanguage) < preferredLanguages.indexOf(newFilename))

        logger().debug("oldFileNeedRename: {}, betterLanguageFound: {}, newFileIsValidName: {}", oldFileNeedRename, betterLanguageFound, newFileIsValidName)

        return (oldFileNeedRename || betterLanguageFound) && newFileIsValidName
    }

    private fun getNextFile(filename: String): File {
        val cleanFilename = filename.replace("/", ", ")
        var file = File(folderName + File.separator + cleanFilename + ".jpg")
        var fileNo = 0
        while (file.exists()) {
            fileNo++
            file = File(folderName + File.separator + cleanFilename + fileNo + ".jpg")
        }
        return file
    }
}
